### PR TITLE
New custom games and custom rules screen

### PR DESCRIPTION
**Pull request by Natchuz**
This includes my commits from my last pull requests (I've closed it), cause I've made a branch in wrong place, don't blame me for that ;_;
\*\*BUT\*\*
this AwEsOmE pull request introduces:

- Completely new custom games screen.
- Now you set custom rules by modifying selected gamemode
- Finally no mutually exclusive options.
- Renamed RulePreset to Gamemode (cause they have different goals)
- Removed waves from sandbox (Now it's only freebuild mode)
- And new rules from last pull request: new respawning and waves options

### CODE REVIEW

File: core/src/io/anuke/mindustry/ui/dialogs/CustomGameDialog.java [hunk 1]
```diff
@@ -155,4 +158,89 @@ private void displayGameModeHelp(){
         d.show();
     }
 
-}
+    private class CustomRulesDialog extends FloatingDialog{
```
Comment (Anuken) [inline core/src/io/anuke/mindustry/ui/dialogs/CustomGameDialog.java]: This should not be an inner class. It's may be related, but it's still a separate dialog and a rather large class file.

File: core/src/io/anuke/mindustry/ui/dialogs/CustomGameDialog.java [hunk 2]
```diff
@@ -155,4 +158,89 @@ private void displayGameModeHelp(){
...
+        }
+
+        void title(String text, Gamemode... gamemodes) {
+            if(Arrays.asList(gamemodes).contains(selectedGamemode)) {
+                title(text);
+            }
+        }
+
+        void title(String text) {
+            main.add(text).color(Color.CORAL).fontScale(1.5f).padTop(40).padBottom(20).padRight(100f);
```
Comment (Anuken) [inline core/src/io/anuke/mindustry/ui/dialogs/CustomGameDialog.java]: Use `Pal.accent`; I would like to avoid extra colors.

File: core/src/io/anuke/mindustry/ui/dialogs/CustomGameDialog.java [hunk 3]
```diff
@@ -155,4 +158,89 @@ private void displayGameModeHelp(){
...
+        }
+
+        void check(String text, BooleanConsumer cons, BooleanProvider prov, BooleanProvider condition, Gamemode... gamemodes) {
+            if(Arrays.asList(gamemodes).contains(selectedGamemode)) {
+                check(text, cons, prov, condition);
+            }
+        }
+
+        void title(String text, Gamemode... gamemodes) {
+            if(Arrays.asList(gamemodes).contains(selectedGamemode)) {
```
Comment (Anuken) [inline core/src/io/anuke/mindustry/ui/dialogs/CustomGameDialog.java]: Use `Structs.contains(gamemodes, selectedGamemode)`.

File: core/src/io/anuke/mindustry/ui/dialogs/CustomGameDialog.java [hunk 4]
```diff
@@ -155,4 +158,89 @@ private void displayGameModeHelp(){
...
+            }
+        }
+
+        void check(String text, BooleanConsumer cons, BooleanProvider prov, BooleanProvider condition) {
+            main.addCheck(text, cons).checked(prov.get()).update(a -> a.setDisabled(!condition.get())).padRight(100f);
+            main.row();
+        }
+
+        void check(String text, BooleanConsumer cons, BooleanProvider prov, BooleanProvider condition, Gamemode... gamemodes) {
+            if(Arrays.asList(gamemodes).contains(selectedGamemode)) {
```
Comment (Anuken) [inline core/src/io/anuke/mindustry/ui/dialogs/CustomGameDialog.java]: See comment below.

File: core/src/io/anuke/mindustry/ui/dialogs/CustomGameDialog.java [hunk 5]
```diff
@@ -155,4 +158,89 @@ private void displayGameModeHelp(){
...
+            setFillParent(true);
+            shown(this::setup);
+            addCloseButton();
+        }
+
+        void setup(){
+            cont.clear();
+            cont.pane(m -> main = m);
+            main.margin(10f);
+            main.setWidth(1000f);
```
Comment (Anuken) [inline core/src/io/anuke/mindustry/ui/dialogs/CustomGameDialog.java]: Setting the width of an element that participates in layout doesn't (or shouldn't) do anything. Even if it did, this is far too wide for a mobile device to fit on their screen in portrait mode.

**Comment by Natchuz** [comment]
I'll do it tomorrow

**Comment by Natchuz** [comment]
Have I missed any exclusion?

File: core/src/io/anuke/mindustry/ui/dialogs/CustomGameDialog.java (hunk 1 above)
Comment (Natchuz) [reply core/src/io/anuke/mindustry/ui/dialogs/CustomGameDialog.java]: How it should be related to make it consistent? Should I make a function that will sync vars between classes, or in ``CustomRulesDialog`` should I just access vars by ``ui.custom.``?

**Comment by Natchuz** [comment]
Done. @Anuken

**Comment by Anuken** [comment]
Will test this put when I can.

File: core/assets/bundles/bundle.properties [hunk 6]
```diff
@@ -520,6 +518,17 @@ rules.respawntime = Respawn Time:[LIGHT_GRAY] (sec)
 rules.wavespacing = Wave Spacing:[LIGHT_GRAY] (sec)
 rules.buildcostmultiplier = Build Cost Multiplier
 rules.buildspeedmultiplier = Build Speed Multiplier
+rules.waitForWaveToEnd = Begin counting to next wave after all enemies are killed
+rules.respawns = Number of maximum respawns per wave
```
Comment (Luxray5474) [inline core/assets/bundles/bundle.properties]: (reference 1) 
```suggestion
Max respawns per wave (0 for unlimited)
```

File: core/assets/bundles/bundle.properties [hunk 7]
```diff
@@ -520,6 +518,17 @@ rules.respawntime = Respawn Time:[LIGHT_GRAY] (sec)
...
+rules.limitedRespawns = Should respawns be limted?
```
Comment (Luxray5474) [inline core/assets/bundles/bundle.properties]: (refer to comment on line 522) I don't think you should have a button just for this. I would say it's simpler to have just a number input and assign 0 to unlimited.

File: core/assets/bundles/bundle.properties [hunk 8]
```diff
@@ -495,20 +496,17 @@ mode.help.title = Description of modes
...
+mode.attack.description = Destroy enemy's base.
 mode.custom = Custom Rules
 
 rules.infiniteresources = Infinite Resources
 rules.wavetimer = Wave Timer
 rules.waves = Waves
 rules.enemyCheat = Infinite AI (Red Team) Resources
-rules.pvp = PvP
 rules.unitdrops = Unit Drops
 rules.unitbuildspeedmultiplier = Unit Creation Speed Multiplier
```
Comment (Luxray5474) [inline core/assets/bundles/bundle.properties]:
```suggestion
Unit Production Speed Multiplier
```

File: core/assets/bundles/bundle.properties [hunk 9]
```diff
@@ -520,6 +518,17 @@ rules.respawntime = Respawn Time:[LIGHT_GRAY] (sec)
 rules.wavespacing = Wave Spacing:[LIGHT_GRAY] (sec)
 rules.buildcostmultiplier = Build Cost Multiplier
 rules.buildspeedmultiplier = Build Speed Multiplier
+rules.waitForWaveToEnd = Begin counting to next wave after all enemies are killed
```
Comment (Luxray5474) [inline core/assets/bundles/bundle.properties]:
```suggestion
Count Down to Next Wave
```

File: core/src/io/anuke/mindustry/entities/type/Player.java [hunk 10]
```diff
@@ -490,6 +494,10 @@ public void drawBuildRequests(){
     @Override
     public void update(){
 
+        if((unitGroups[waveTeam.ordinal()].size() == 0) && state.rules.limitedRespawns){
```
Comment (Luxray5474) [inline core/src/io/anuke/mindustry/entities/type/Player.java]: (refers to comment on line 522, 523)
```suggestion
if((unitGroups[waveTeam.ordinal()].size() == 0) && state.rules.respawns != 0){
```

File: core/src/io/anuke/mindustry/entities/type/Player.java [hunk 11]
```diff
@@ -103,6 +104,9 @@ public void hitboxTile(Rectangle rectangle){
     public void onRespawn(Tile tile){
         boostHeat = 1f;
         achievedFlight = true;
+        if(state.rules.limitedRespawns){
```
Comment (Luxray5474) [inline core/src/io/anuke/mindustry/entities/type/Player.java]: (refers to comment on line 522, 523)

https://github.com/Anuken/Mindustry/blob/b9421f8c5ab6ecdbf1b1ddf43ad592abc1388911/core/src/io/anuke/mindustry/entities/type/Player.java#L107-L109

```suggestion
if(state.rules.respawns == 0){
  respawns--;
}
```

File: core/src/io/anuke/mindustry/game/Rules.java [hunk 12]
```diff
@@ -45,4 +50,10 @@
     public byte zone = -1;
     /** Spawn layout. Since only zones modify this, it should be assigned on save load. */
     public transient Array<SpawnGroup> spawns = DefaultWaves.get();
+    /** Determines if there should be limited respawns. */
+    public boolean limitedRespawns = false;
```
Comment (Luxray5474) [inline core/src/io/anuke/mindustry/game/Rules.java]: (refers to comment on line 522, 523)
```suggestion
```

File: core/src/io/anuke/mindustry/ui/dialogs/CustomRulesDialog.java [hunk 13]
```diff
@@ -1,67 +1,106 @@
...
-        number("$rules.enemycorebuildradius", f -> rules.enemyCoreBuildRadius = f * tilesize, () -> Math.min(rules.enemyCoreBuildRadius / tilesize, 200));
-        number("$rules.respawntime", f -> rules.respawnTime = f * 60f, () -> rules.respawnTime / 60f);
-        number("$rules.wavespacing", f -> rules.waveSpacing = f * 60f, () -> rules.waveSpacing / 60f);
+        title("$rules.title.waves", Gamemode.survival);
+        check("$rules.waves", b -> rules.waves = b, () -> rules.waves, ()->false, Gamemode.survival);
+        check("$rules.wavetimer", b -> rules.waveTimer = b, () -> rules.waveTimer, ()->rules.waves, Gamemode.survival);
+        check("$rules.waitForWaveToEnd", b -> rules.waitForWaveToEnd = b, () -> rules.waitForWaveToEnd, ()->rules.waves, Gamemode.survival);
+        number("$rules.wavespacing", f -> rules.waveSpacing = f * 60f, () -> rules.waveSpacing / 60f, ()->rules.waves, Gamemode.survival);
+        title("$rules.title.respawns");
+        check("$rules.limitedRespawns", b -> rules.limitedRespawns= b, () -> rules.limitedRespawns, ()->true);
```
Comment (Luxray5474) [inline core/src/io/anuke/mindustry/ui/dialogs/CustomRulesDialog.java]: (refers to comment on line 522, 523)
```suggestion
```

**Comment by Natchuz** [comment]
@Luxray5474 The point is that someone would like to have 0 respawns per wave. Also, if we can have system that protects from mutually exclusives options, we should use it. It's simpler for users.

**Comment by Luxray5474** [comment]
I see. I didn't think of that, my mistake.

File: core/assets/bundles/bundle.properties [hunk 14]
```diff
@@ -495,20 +496,17 @@ mode.help.title = Description of modes
 mode.survival.name = Survival
 mode.survival.description = The normal mode. Limited resources and automatic incoming waves.
 mode.sandbox.name = Sandbox
-mode.sandbox.description = Infinite resources and no timer for waves.
-mode.freebuild.name = Freebuild
-mode.freebuild.description = Limited resources and no timer for waves.
+mode.sandbox.description = Build, build, build, repeat. No waves, no stress.
```
[the remark this answers is not in the document]
Comment (Anuken) [reply core/assets/bundles/bundle.properties]: I don't think either of these descriptions are ~~depressing enough~~ fit the game's tone.

[the remark this answers is not in the document]
Comment (Natchuz) [reply core/assets/bundles/bundle.properties]: :thinking: 🤔

File: core/assets/bundles/bundle.properties (hunk 9 above)
Comment (Natchuz) [reply core/assets/bundles/bundle.properties]: I guess this one ain't accurate

Comment (Natchuz) [reply core/assets/bundles/bundle.properties]: Count down to next wave as all enemies are destroyed

2 chars shorter. Maybe better

File: core/assets/bundles/bundle.properties (hunk 14 above)
[the remark this answers is not in the document]
Comment (Natchuz) [reply core/assets/bundles/bundle.properties]: But I liked that one

**Comment by Luxray5474** [comment]
Uhh..... https://i.imgur.com/wiRof0v.png

**Comment by Natchuz** [comment]
🤔 
Also, my game started to crash if I don't open "modify rules" dialog before I launch the map. Didn't happen before

File: core/assets/bundles/bundle.properties (hunk 14 above)
[the remark this answers is not in the document]
Comment (Luxray5474) [reply core/assets/bundles/bundle.properties]: I guess `repeat` only fits with like `eat sleep <verb> repeat` like `eat sleep build repeat` but not with the same items. ~~also i observe `repeat` carries onto 'eat'~~

File: core/assets/bundles/bundle.properties (hunk 9 above)
Comment (Luxray5474) [reply core/assets/bundles/bundle.properties]: So basically it does not count down after the wave has fired but instead after all enemies are killed, is what I am getting here.

`Start wave timer when all enemies are destroyed` fits better because `as` suggests `while` all the enemies are being destroyed.

**Comment by Luxray5474** [comment]
Doesn't happen to me.

**Comment by Natchuz** [comment]
Weird

**Comment by Luxray5474** [comment]
Also, I found that disabling `Waves` but having `Wave Timer` on suggests the Attack mode? It would be nice if Anuke implemented a `one up one down` sort of system where pressing a button disables others. It's weird that you can enable wave timers but not waves themselves.

**Comment by Natchuz** [comment]
What do you mean by that? I mean, survival is made for waves, and you can't disable these in this mode, but you can choose whether ``Wave Timer`` should be enabled. Other modes don't have the option to enable waves. This is what I did. Disabling ``Waves`` disables also ``Wave Timer``.

**Comment by Natchuz** [comment]
Oh yea

**Comment by Natchuz** [comment]
So now, does this crash happens to you?

**Comment by Natchuz** [comment]
My branches are mess

**Comment by Luxray5474** [comment]
Yes.

```
java.lang.NullPointerException
        at io.anuke.mindustry.game.Teams.isActive(Teams.java:31)
        at io.anuke.mindustry.ai.Pathfinder.clear(Pathfinder.java:188)
        at io.anuke.mindustry.ai.Pathfinder.lambda$new$0(Pathfinder.java:28)
        at io.anuke.arc.Events.fire(Events.java:24)
        at io.anuke.mindustry.core.World.endMapLoad(World.java:182)
        at io.anuke.mindustry.core.World.loadMap(World.java:248)
        at io.anuke.mindustry.core.Control.lambda$playMap$16(Control.java:202)
        at io.anuke.mindustry.core.UI.lambda$loadAnd$4(UI.java:205)
        at io.anuke.arc.util.Time$1.run(Time.java:29)
        at io.anuke.arc.backends.lwjgl3.Lwjgl3Application.loop(Lwjgl3Application.java:248)
        at io.anuke.arc.backends.lwjgl3.Lwjgl3Application.<init>(Lwjgl3Application.java:59)
        at io.anuke.mindustry.desktop.DesktopLauncher.<init>(DesktopLauncher.java:19)        at io.anuke.mindustry.desktop.DesktopLauncher.main(DesktopLauncher.java:50)
```

**Comment by Natchuz** [comment]
Idk, this never happened before. It has appeared from nowhere

File: core/assets/bundles/bundle.properties (hunk 14 above)
[the remark this answers is not in the document]
Comment (Natchuz) [reply core/assets/bundles/bundle.properties]: @Anuken what do you propose?

**Comment by Natchuz** [comment]
Meh, I'll go to sleep now (EU time). I'll figure this out tomorrow.

**Comment by Natchuz** [comment]
image
I am a git master... If you wonder what those commits are

File: core/assets/bundles/bundle.properties (hunk 14 above)
[the remark this answers is not in the document]
Comment (Anuken) [reply core/assets/bundles/bundle.properties]: Keep the old description, I'll write a new one much later.

**Comment by Natchuz** [comment]
I'm alive AND I fixed this crash but discovered a new bug where I don't know why, but both sandbox and survival require enemy's core.
Working on it

**Comment by Natchuz** [comment]
@Anuken I guess it's finally ready

File: core/src/io/anuke/mindustry/game/Rules.java [hunk 15]
```diff
@@ -45,4 +50,12 @@
...
+    /** How many times player can respawn during one wave. */
+    public int respawns = 5;
+    /** Hold wave timer until all enemies are destroyed. */
+    public boolean waitForWaveToEnd = false;
+    /** Determinates if gamemode requires more than 1 cores to play */
+    public boolean manyCores = false;
```
Comment (Anuken) [inline core/src/io/anuke/mindustry/game/Rules.java]: I don't think this variable name makes sense. What it really controls is the *objective* of the map, not whether it requires multiple cores to play. Something like `attackMode` might be better, but even then I'm not too happy with hardcoded objectives like this. A flexible `Objective` enum would be the best solution, but that's outside the scope of this PR.

File: core/src/io/anuke/mindustry/game/Rules.java [hunk 16]
```diff
@@ -2,6 +2,11 @@
 
 import io.anuke.annotations.Annotations.Serialize;
 import io.anuke.arc.collection.Array;
+import io.anuke.arc.function.BooleanConsumer;
+import io.anuke.arc.function.BooleanProvider;
+import io.anuke.arc.function.Function;
+
+import java.util.function.Predicate;
```
Comment (Anuken) [inline core/src/io/anuke/mindustry/game/Rules.java]: Why are these extra imports here? They're not used in the class. It's also worth mentioning that anything out of `java.util.function` isn't supported on Android; avoid it.

File: core/src/io/anuke/mindustry/ui/dialogs/CustomGameDialog.java [hunk 17]
```diff
@@ -94,7 +93,10 @@ void setup(){
             state.wavetime = difficulty.waveTime;
         }).width(s);
 
-        cont.add(sdif).visible(() -> lastPreset != null);
+        cont.add(sdif);//.visible(() -> lastPreset != null);
```
Comment (Anuken) [inline core/src/io/anuke/mindustry/ui/dialogs/CustomGameDialog.java]: Either remove the `visible` code or keep it altogether, don't comment it out if you don't need it.

File: core/src/io/anuke/mindustry/entities/type/Player.java (hunk 10 above)
Comment (Anuken) [inline core/src/io/anuke/mindustry/entities/type/Player.java]: If respawns are reset after each wave and each call to reset(), why is this necessary?

File: core/src/io/anuke/mindustry/ui/dialogs/CustomRulesDialog.java [hunk 18]
```diff
@@ -1,67 +1,106 @@
...
-        number("$rules.buildspeedmultiplier", f -> rules.buildSpeedMultiplier = f, () -> rules.buildSpeedMultiplier);
-        number("$rules.unitbuildspeedmultiplier", f -> rules.unitBuildSpeedMultiplier = f, () -> rules.unitBuildSpeedMultiplier);
-        number("$rules.unithealthmultiplier", f -> rules.unitHealthMultiplier = f, () -> rules.unitHealthMultiplier);
-        number("$rules.playerhealthmultiplier", f -> rules.playerHealthMultiplier = f, () -> rules.playerHealthMultiplier);
-        number("$rules.playerdamagemultiplier", f -> rules.playerDamageMultiplier = f, () -> rules.playerDamageMultiplier);
-        number("$rules.unitdamagemultiplier", f -> rules.unitDamageMultiplier = f, () -> rules.unitDamageMultiplier);
-        number("$rules.enemycorebuildradius", f -> rules.enemyCoreBuildRadius = f * tilesize, () -> Math.min(rules.enemyCoreBuildRadius / tilesize, 200));
-        number("$rules.respawntime", f -> rules.respawnTime = f * 60f, () -> rules.respawnTime / 60f);
-        number("$rules.wavespacing", f -> rules.waveSpacing = f * 60f, () -> rules.waveSpacing / 60f);
+        title("$rules.title.waves", Gamemode.survival);
```
Comment (Anuken) [inline core/src/io/anuke/mindustry/ui/dialogs/CustomRulesDialog.java]: There's no way to configure `manyCores` here. This makes it impossible to play an attack-based custom gamemode.

File: core/src/io/anuke/mindustry/world/blocks/storage/CoreBlock.java [hunk 19]
```diff
@@ -150,8 +150,9 @@ public void update(Tile tile){
 
             entity.heat = Mathf.lerpDelta(entity.heat, 1f, 0.1f);
             entity.time += entity.delta();
-            entity.progress += 1f / state.rules.respawnTime * entity.delta();
-
+            if(entity.currentUnit.respawns > 0){
```
Comment (Anuken) [inline core/src/io/anuke/mindustry/world/blocks/storage/CoreBlock.java]: First of all, players can still respawn in `MechPad`s with 0 respawns. Second of all, I don't think the respawning blocks should handle this logic-- players should simply not be placed in respawn points when their respawns run out. This can be done by not calling `Player#updateRespawning` when the player has <= 0 respawns.

File: core/src/io/anuke/mindustry/game/Rules.java (hunk 16 above)
Comment (Natchuz) [reply core/src/io/anuke/mindustry/game/Rules.java]: hmm... I just forgot to cleanup imports

File: core/src/io/anuke/mindustry/ui/dialogs/CustomGameDialog.java (hunk 17 above)
Comment (Natchuz) [reply core/src/io/anuke/mindustry/ui/dialogs/CustomGameDialog.java]: Oh sure, that's an omission.

File: core/src/io/anuke/mindustry/ui/dialogs/CustomRulesDialog.java (hunk 18 above)
Comment (Natchuz) [reply core/src/io/anuke/mindustry/ui/dialogs/CustomRulesDialog.java]: But wait. It's set by default when you select Attack gamemode. New custom-rules screen only allows you to modify rules which don't effect objective. That's why RulePreset is now Gamemode (they have different goals)

File: core/src/io/anuke/mindustry/game/Rules.java (hunk 15 above)
Comment (Natchuz) [reply core/src/io/anuke/mindustry/game/Rules.java]: `attackMode` is a perfect solution for now. `Objective` enum would be great, but also `Gamemode` is kind of `Objective` itself. How I mentioned below, that's why `Gamemode` is no longer only `RulePreset `

File: core/src/io/anuke/mindustry/entities/type/Player.java (hunk 10 above)
Comment (Natchuz) [reply core/src/io/anuke/mindustry/entities/type/Player.java]: Sure

File: core/src/io/anuke/mindustry/game/Rules.java (hunk 15 above)
Comment (Natchuz) [reply core/src/io/anuke/mindustry/game/Rules.java]: Following this way, we can also delete PVP rule, and handle logic by checking game's gamemode instead of single rule. Maybe that's the way.

Comment (Natchuz) [reply core/src/io/anuke/mindustry/game/Rules.java]: But this kind of rework would be useful only if we want to introduce new mixed modes. For now, I guess it's sufficient.

**Comment by Natchuz** [comment]
@Anuken Done. A couple of explanations in the review's comments. Your all doubts about custom gamemode creating mechanics will be resolved when you'll test this out.

**Comment by Anuken** [comment]
*Some issues I found with the UI.*

---
Screenshot from 2019-04-18 17-14-18

Don't change the font scale! Mindustry uses bitmap fonts, which don't scale well. Use an underline if you want to divide sections.

---
Screenshot from 2019-04-18 17-16-21

The "modify rules" button doesn't have consistent capitalizaton (should be "Modify Rules") and takes up too much space. This is a significant problem if you're on a small phone in landscape. Try moving it into the same line as the difficulty selection table. To save even more space, the text can be changed to one word, e.g. "Customize". *I may need to redesign the entire custom game dialog at some point.*

---
Screenshot from 2019-04-18 17-18-40

When on mobile in portrait mode, the gamemodes get cut off. That's why they were on separate lines originally.
*(This is a desktop demonstration, but it would look similar on a real phone.)*

---
Screenshot from 2019-04-18 17-21-53

This description doesn't sound quite right, and seems inconsistent with other descriptions. Revert it to the old description for now.

---
Screenshot from 2019-04-18 17-23-57

Why not use the already-present "reset to defaults" text? Capitalization is also inconsistent here.

---
Screenshot from 2019-04-18 17-25-25

Padding for the title and elements seems a bit too high. Might cause usability issues for phones in landscape.

---
Screenshot from 2019-04-18 17-27-16

This description is too verbose and may not fit on portrait screens. How about "Waves wait for enemies" or something similar?

File: core/src/io/anuke/mindustry/ui/dialogs/CustomRulesDialog.java [hunk 20]
```diff
@@ -1,67 +1,105 @@
...
 
-        rules.waves = true;
-        rules.waveTimer = true;
-
         setFillParent(true);
         shown(this::setup);
         addCloseButton();
     }
 
+    public void show(Rules rules, Gamemode gamemode){this.rules = rules; this.selectedGamemode = gamemode; show();}
```
Comment (Anuken) [inline core/src/io/anuke/mindustry/ui/dialogs/CustomRulesDialog.java]: Forgot to mention: These statements should be in separate lines. I only sent an example in Discord to give a general idea of how it would work.

**Comment by Natchuz** [comment]
image
@Anuken This is what I got. I there any other way to do it?

**Comment by Natchuz** [comment]
Leaving it as it is. If there is anything else to change, I'll do it tomorrow afternoon of EU time

File: core/src/io/anuke/mindustry/game/Gamemode.java [hunk 21]
```diff
@@ -13,14 +13,15 @@
     }}),
     sandbox(() -> new Rules(){{
         infiniteResources = true;
-        waves = true;
+        waves = false;
```
Comment (Anuken) [inline core/src/io/anuke/mindustry/game/Gamemode.java]: Why disable waves for sandbox? I realize that it's supposed to be "safe" with no waves, but considering players can control waves, there's no reason to disable them outright. This will just cause confusion; people will play sandbox expecting a button to run waves, not knowing that you need to go and configure custom game options.

Comment (Natchuz) [reply core/src/io/anuke/mindustry/game/Gamemode.java]: @Anuken Yes, I understand your way of thinking. But you can disable wave timer and enable infinite resources in survival mode. I wanted to make sandbox a freebuild mode with no objective. And survival's objective is to survive waves. Same like the old sandbox. If you wish I can do that your way, but I propose renaming sandbox to freebuild which would let peoples know that something changed. Let me know what do you think of that.

Comment (Anuken) [reply core/src/io/anuke/mindustry/game/Gamemode.java]: > But you can disable wave timer and enable infinite resources in survival mode.
You can, but people who want waves in sandbox won't know that. Considering many people can't even find the "custom game" button, I can already see people asking to have waves back in sandbox.

> I wanted to make sandbox a freebuild mode with no objective.
Why? I don't see any reason to do this. Having a button to press which spawns waves in no way hurts the experience. In fact, many people playing sandbox are probably doing it to test the effectiveness of turrets against enemies, which is impossible in this modified sandbox mode. Every time someone would like to test waves, they would need to go in and open a new dialog to do so. This is not a good user experience.

> but I propose renaming sandbox to freebuild which would let peoples know that something changed.
If I keep waves in sandbox, nothing would have changed. It would work the same way it does in 3.5. Even then, that's *not* equivalent to the old freebuild. This would just cause confusion.

**Comment by Natchuz** [comment]
Ok, you convinced me. Should I put wave timer and other waves buttons in custom rules screen when selected sandbox gamemode then?

File: core/src/io/anuke/mindustry/game/Gamemode.java (hunk 21 above)
Comment (Natchuz) [reply core/src/io/anuke/mindustry/game/Gamemode.java]: Ok, you convinced me. Should I put wave timer and other waves buttons in custom rules screen when selected sandbox gamemode then?

Comment (Anuken) [reply core/src/io/anuke/mindustry/game/Gamemode.java]: Yes.

**Comment by Natchuz** [comment]
@Anuken done

File: core/src/io/anuke/mindustry/ui/dialogs/CustomRulesDialog.java [hunk 22]
```diff
@@ -1,67 +1,128 @@
...
-            .valid(Strings::canParsePositiveFloat).width(120f).left().get());
-        });
+                    .padRight(100f)
+                    .update(a -> a.setDisabled(!condition.get()))
+                    .valid(Strings::canParsePositiveFloat).width(120f) .left().get());
+        }).padTop(0);
+        main.row();
+    }
+
+    void f_number(String text, FloatConsumer cons, FloatProvider prov, BooleanProvider condition, Gamemode... gamemodes){
```
Comment (Anuken) [inline core/src/io/anuke/mindustry/ui/dialogs/CustomRulesDialog.java]: One last thing: In Java conventions, method names shouldn't have underscores. Instead of `f_number` and `i_number`, use `number` (for floats, since that's most common) and `integer`.

**Comment by Natchuz** [comment]
@Anuken Done.

**Comment by Anuken** [comment]
Alright, looks good, thanks!
